### PR TITLE
Add LSP tracing

### PR DESCRIPTION
**Pull request by AleksanderBG**
LSP communication can now be traced remotely.

See https://github.com/lampepfl/dotty-remote-tracer for server implementation.

### CODE REVIEW

File: vscode-dotty/src/tracer.ts [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
+    createLspOutputChannel(): vscode.OutputChannel | undefined {
+        const remoteTracingUrl = this.ctx.extensionConfig.get<string>('remoteTracingUrl');
+        if (!remoteTracingUrl) return undefined;
+
+        if (this.tracingConsent.get() === 'no-answer') {
+            vscode.window.showInformationMessage(
+                'Do you want to help EPFL develop this plugin by uploading your usage data? ' +
+                'PLEASE BE AWARE that this will upload all of your keystrokes and all of your code, ' +
+                'among other things.',
+                'yes', 'no'
```
Comment (Duhemm) [inline vscode-dotty/src/tracer.ts]: I think that just saying "among other things" is scary. We should either exhaustively list what's being uploaded here or include a link to a page that exhaustively lists what's being uploaded.

Also, I would replace "all of your keystrokes" by "every action performed in Dotty IDE". Keystrokes outside of the IDE are not recorded.

**Comment by AleksanderBG** [comment]
@Duhemm fixed most of the things you mentioned. Tracing popup now directly mentions that we simply upload all LSP communication, which contains all the files and all the keystrokes. Added a statusbar item which tells the user tracing is on and allows them to adjust their consent.

There are still some outlying issues (Dotty codebase cannot be really traced, b/c the files it contains are too large; websocket connections get dropped if internet connection disappers for a second and cannot be recreated because the server does not support it; we should probably clean the LSP output channel from time to time to avoid too much trash, but do so in a way consistent with other plugins), but this was meant to work for students and it should be sufficient for them. @smarter?

**Comment by smarter** [comment]
> Dotty codebase cannot be really traced, b/c the files it contains are too large

What's the exact issue here? Can we work around it by sending compressed files ?

**Comment by AleksanderBG** [comment]
@smarter I think I found a good solution. The problem starts with the fact that LSP sometimes sends entire files in one message, at least when the file is opened. This causes issues, since there's a limit on the size of a message sent through WebSocket (probably both in Play Framework and Nginx), and it may not be possible (or desireable) to completely remove it. Allowing arbitrary size messages may open the server to easy DoS, and setting a limit on the size of message will ultimately just delay the problem.

The solution would be to send messages in chunks if they are too large, and save each chunk to disk immediately after it is received. This avoid the problem of too large messages, but creates a possibility where the last line of the log on the disk is malformed (because not all chunks were sent/received). I think we can work around that limitation more easily, since it's enough to just check if the file ends in a newline.

**Comment by AleksanderBG** [comment]
Rebased, updated `npm` to a version which actually notices and uses the `package-lock.json`, added a setting for maximum message size. Could not get the connection to fail while opening files in Dotty repo.

File: vscode-dotty/src/tracer.ts [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+            }
+
+            return persisted(uuidv4())
+        })()
+
+        this.projectId = vscode.workspace.name !== undefined ? vscode.workspace.name : 'no-project'
+        this.sessionId = new Date().toISOString()
+    }
+
+    run(): { lspOutputChannel?: vscode.OutputChannel } {
```
Comment (smarter) [inline vscode-dotty/src/tracer.ts]: Why not `run(): vscode.OutputChannel | undefined` ?

File: vscode-dotty/src/tracer.ts [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+            'including every single keystroke.',
+            'yes', 'no'
+        ).then((value: string | undefined) => {
+            if (value === 'yes' || value === 'no') this.tracingConsent.set(value)
+        })
+    }
+
+    private initializeAsyncWorkspaceDump() {
+        if (this.remoteWorkspaceDumpUrl === undefined) return
+        // convince TS that this is a string
```
Comment (smarter) [inline vscode-dotty/src/tracer.ts]: You can convince TS of this in a more direct way:
```typescript
const url = this.remoteWorkspaceDumpUrl
if (url) {
  // ...
}
```

File: vscode-dotty/src/tracer.ts [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+                this.logErrorWithoutNotifying(
+                    'socket error',
+                    remoteTracingUrl,
+                    new SafeJsonifier(event, (event) => ({
+                        error: safeError(event.error),
+                        message: event.message,
+                        type: event.type
+                    }))
+                )
+                vscode.window.showWarningMessage('An error occured in Dotty LSP remote tracing connection.')
```
Comment (smarter) [inline vscode-dotty/src/tracer.ts]: I would rather not show any messages to users if something goes wrong, it's distracting and not very useful for them. What we could do is change the status bar message to have On/Off/Failed

File: vscode-dotty/src/tracer.ts [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+function safeError(e: Error): SafeJsonifier<Error> {
+    return new SafeJsonifier(e, (e) => e.toString())
+}
+
+class SafeJsonifier<T> {
+    value: T
+    valueToObject: (t: T) => {}
+
+    constructor(value: T, valueToObject: (t: T) => {}) {
```
Comment (smarter) [inline vscode-dotty/src/tracer.ts]: You can directly create an immutable class field and assign by writing `constructor(readonly value: T, ...)`

File: vscode-dotty/src/tracer.ts [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+            })
+        }
+    }
+}
+
+function safeError(e: Error): SafeJsonifier<Error> {
+    return new SafeJsonifier(e, (e) => e.toString())
+}
+
+class SafeJsonifier<T> {
```
Comment (smarter) [inline vscode-dotty/src/tracer.ts]: Some documentation would be nice :)

File: vscode-dotty/src/tracer.ts [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+    private get isTracingEnabled(): boolean {
+        return Boolean(this.remoteWorkspaceDumpUrl || this.remoteTracingUrl)
+    }
+
+    constructor(ctx: Ctx) {
+        this.ctx = ctx
+
+        this.tracingConsent = new TracingConsentCache(ctx.extensionContext.workspaceState)
+
+        this.remoteWorkspaceDumpUrl = this.ctx.extensionConfig.get<string>('remoteWorkspaceDumpUrl')
```
Comment (smarter) [inline vscode-dotty/src/tracer.ts]: I think these should be in "tracing." like maximumMessageSize

File: vscode-dotty/src/tracer.ts [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+            })()
+        }
+        renderStatusBarItem()
+        this.tracingConsent.subscribe(renderStatusBarItem)
+        item.show()
+        return item
+    }
+
+    private createLspOutputChannel(): vscode.OutputChannel | undefined {
+        if (!this.remoteTracingUrl) return undefined
```
Comment (smarter) [inline vscode-dotty/src/tracer.ts]: Unless I'm missing something, it looks like this means that we don't get any outputchannel, even a local one, when remoteTracingUrl is undefined. Suggestion: make an LspOutputChannel that extends OutputChannel and take as argument a (local) OutputChannel, then every method of LspOutputChannel just forwards to the underlying OutputChannel and also does network stuff if tracing is enabled.

**Pull request merged**